### ISSUE TITLE
Unable to install 0.46 with npm on Windows

### ISSUE THREAD

**Issue by SanokKule** [rejetto/hfs#285]
Trying to install npm version of hfs 0.46 on Windows will fail deleting `node_modules\hfs\plugins` and `node_modules\hfs\src` which don't exist, and requires Visual Studio to compile fs-extended-attributes
```
PS C:\Users\SanokKule> npm install hfs@latest
npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     'C:\\Users\\SanokKule\\node_modules',
npm WARN cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\SanokKule\node_modules\hfs\plugins'] {
npm WARN cleanup       errno: -4048,
npm WARN cleanup       code: 'EPERM',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: 'C:\\Users\\SanokKule\\node_modules\\hfs\\plugins'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
npm ERR! code 1
npm ERR! path C:\Users\SanokKule\node_modules\fs-extended-attributes
npm ERR! command failed
npm ERR! command C:\Windows\system32\cmd.exe /d /s /c node-gyp-build
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@20.1.0 | win32 | x64
npm ERR! gyp info find Python using Python version 3.11.2 found at "C:\Program Files\Python311\python.exe"
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! find VS msvs_version not set from command line or npm config
npm ERR! gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
npm ERR! gyp ERR! find VS could not use PowerShell to find Visual Studio 2017 or newer, try re-running with '--loglevel silly' for more details
npm ERR! gyp ERR! find VS not looking for VS2015 as it is only supported up to Node.js 18
npm ERR! gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS You need to install the latest version of Visual Studio
npm ERR! gyp ERR! find VS including the "Desktop development with C++" workload.
npm ERR! gyp ERR! find VS For more information consult the documentation at:
npm ERR! gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
npm ERR! gyp ERR! find VS **************************************************************
npm ERR! gyp ERR! find VS
npm ERR! gyp ERR! configure error
npm ERR! gyp ERR! stack Error: Could not find any Visual Studio installation to use
npm ERR! gyp ERR! stack     at VisualStudioFinder.fail (C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:122:47)
npm ERR! gyp ERR! stack     at C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:75:16
npm ERR! gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:380:14)
npm ERR! gyp ERR! stack     at C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:71:14
npm ERR! gyp ERR! stack     at VisualStudioFinder.findVisualStudio2015 (C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:364:14)
npm ERR! gyp ERR! stack     at C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:67:12
npm ERR! gyp ERR! stack     at failPowershell (C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:156:7)
npm ERR! gyp ERR! stack     at VisualStudioFinder.parseData (C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:170:14)
npm ERR! gyp ERR! stack     at C:\Users\SanokKule\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\find-visualstudio.js:143:14
npm ERR! gyp ERR! stack     at ChildProcess.exithandler (node:child_process:411:7)
npm ERR! gyp ERR! System Windows_NT 10.0.22621
npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\SanokKule\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd C:\Users\SanokKule\node_modules\fs-extended-attributes
npm ERR! gyp ERR! node -v v20.1.0
npm ERR! gyp ERR! node-gyp -v v9.4.0
npm ERR! gyp ERR! not ok
```

**Comment by SanokKule** [comment]
After deleting the `node_modules` folder i can't seem to replicate the EPERM error with npm.
But regarding fs-extended-attributes, would [fs-ads](https://github.com/juncaixinchi/fs-ads) work as a replacement on windows?

**Comment by SanokKule** [comment]
npx fails without showing gyp error, sometimes showing:
```
npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     'C:\\Users\\SanokKule\\AppData\\Local\\npm-cache\\_npx\\95660a09cc333978\\node_modules\\hfs',
npm WARN cleanup     [Error: EPERM: operation not permitted, rmdir 'C:\Users\SanokKule\AppData\Local\npm-cache\_npx\95660a09cc333978\node_modules\hfs\plugins'] {
npm WARN cleanup       errno: -4048,
npm WARN cleanup       code: 'EPERM',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: 'C:\\Users\\SanokKule\\AppData\\Local\\npm-cache\\_npx\\95660a09cc333978\\node_modules\\hfs\\plugins'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
```
The folder it tries to delete will randomly change each time

**Comment by rejetto** [comment]
> npx fails without showing gyp error, sometimes showing:
> npm WARN cleanup Failed to remove some directories [

this is a warning, so nothing critical.
you said you could remove the folder with same user, so the eperm must be caused by something else, like the folder is not empty or the folder is being used.
Could it be that you had hfs running while also installing new version?

**Comment by rejetto** [comment]
> npm ERR! gyp ERR! find VS **************************************************************
> npm ERR! gyp ERR! find VS You need to install the latest version of Visual Studio
> npm ERR! gyp ERR! find VS including the "Desktop development with C++" workload.
> npm ERR! gyp ERR! find VS For more information consult the documentation at:
> npm ERR! gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
> npm ERR! gyp ERR! find VS **************************************************************

so, this seems to be your error.
I didn't have same problem because I have it already installed.
For the time being, npm/npx version on requires some c++ compiler installed (one that is recognized).
I would really want to get rid of this requirement, caused by a [new library i'm using in 0.46 ](https://github.com/mafintosh/fs-extended-attributes).
I'm using its funcitonality for the "uploader" information, but I planned to use it for other things as well.

it’s very convenient, because if you store information (like "uploader") differently, it gets lost when you move/rename the file, and also the information it’s not deleted if you delete the file. This library solves all these problems.

At the moment I'm not sure how this can be solved, but I will try. Suggestions are welcome, possibly good ones 🫠

**Comment by SanokKule** [comment]
> At the moment I'm not sure how this can be solved, but I will try. Suggestions are welcome, possibly good ones 🫠

> Would [fs-ads](https://github.com/juncaixinchi/fs-ads) work as a replacement on windows?

**Comment by rejetto** [comment]
not exactly: that works only on Windows.
It could solve the problem *on Windows* but only if I find a way to [avoid installing the other library on Windows](https://stackoverflow.com/questions/15176082/npm-package-json-os-specific-dependency).
A have an idea to test when I'll have time: to fork the lib in a way that disables compiling and bundles a few precompiled platforms.

**Comment by rejetto** [comment]
do you get errors if you try to `npm install --no-save fs-xattr` ?

**Comment by SanokKule** [comment]
```
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for fs-xattr@0.4.0: wanted {"os":"!win32"} (current: {"os":"win32"})
npm ERR! notsup Valid os:  !win32
npm ERR! notsup Actual os: win32
```

**Comment by rejetto** [comment]
ok, this shouldn't raise any error instead
`npm i --no-save fs-x-attributes`

**Comment by rejetto** [comment]
or better, `npx hfs@exp`
exp is for experimental

**Comment by SanokKule** [comment]
Works without problems. 👌

**Comment by rejetto** [comment]
this should be fixed by 0.46.1
